### ISSUE TITLE
An addon version ? 

### ISSUE THREAD

**Issue by rayzox57** [luttje/gmod-minigolf#1]
**Is your feature request related to a problem? Please describe.**
My request is not for a bug fix, but rather a port to various gamemodes in order to get the most out of golf! :)

**Describe the solution you'd like**
The solution would be to turn most of the work done on a gamemode into an addon

**Describe alternatives you've considered**
The alternative for the moment would be to use the gamemode, but being a sandbox creator I find it such a shame not to be able to have this addon on my server :( especially since I already have holes available for my map

**Additional context**
I am therefore not asking you to create a tool to add "entities" used in golf, just to transform it into an addon where we could just play with the club and the ball on holes that we would have made on a hammer. I might even be willing to pay for you to port ;) Because a golf addon is really the last thing I'm looking for! ^^

My Steam Account : https://steamcommunity.com/id/Rayzox57

Thank for reading 👍

**Comment by luttje** [comment]
Hi @rayzox57,

Thanks for your suggestion, it's a good idea!

I was planning to eventually turn this gamemode into an addon, but never got to it. I appreciate your offer to pay for this port, though I'm not currently taking on new paid work.

However I've got some free time coming up (next month) where I may be able to take a swing _(pun intended)_ at creating a golf addon from this just for fun. If and when I finish the addon, you may always choose to donate an amount that you feel its worth.

## Notes
To ease the process of turning this gamemode into an addon I do have some thoughts and questions you may have ideas about. I would love to hear any ideas you have. I would prefer to focus on converting, fixing and reducing what exists now, instead of creating too many new features.

1. **Creating minigolf holes:** Currently minigolf maps need to be built in Hammer. That's worth it for a gamemode, but imo not for an addon. How do you currently add holes to your map? Props?
2. **OOB markers**: I can quite easily allow these entities to be spawned by an admin (or player) and work: hole start, hole end and flag. For the Out of Bounds (OOB) brushes I've got some ideas, but would like to hear what ideas you have on how to place them in-game.
3. **Team system:** Currently there's a custom team system in the gamemode (color, password, kicking, etc.). For an addon I would remove it and just use the default gmod team system. What are your thoughts?
4. **Betting:** I'd opt to remove the betting system that currently exists to reduce the amount of work needed.


Thanks for your ideas. I'll post here next month if I get a chance to work on this.

**Comment by rayzox57** [comment]
Hey! Thank you for your very quick response!

For my part I think that at the beginning the addon should be reserved for the most experienced (i.e.: to use golf you have to modify or create a map to add the elements yourself [as for the gamemode]) and depending on its popularity make it accessible with creation "tools".
But I think that the method with hammer should always be present because for example for the oob of the ball we could always make more complex grounds directly with hammer, than a tool.
This will allow the most "complete" servers (example: dark rp servers with their own map) to know that it is possible from hammer to add golf courses ^^

The betting system is in my opinion (as you said) useless, because people themselves can challenge themselves with point exchanges!

I think the same goes for the team system where people can team up if they want.
Because in my opinion the objective of the porting in addon is precisely to let the players play any hole whenever they want, and that we give them the result at the end when they succeed.

The fact of having a final score on all the holes should be avoided I think and it would be up to the players to do so. It also allows you to greatly reduce the workload ;) In the end the addon would be right: a weapon "the golf club" which interacts with a square entity or the ball will be created.

If it hits a trigger it is considered out and will have to start again and if it hits the "hole" entity it is given these points.

**Comment by rayzox57** [comment]
So we should start with something that is simple but complex for those who want to use it, it will give this unique side where the truly motivated will be able to use it with great satisfaction to say to themselves: it's me who is doing it. 'is made !

And then once the hype is done, and the demands grow for a way to do it without a hammer, update it with a tool system (which will increase its hype even more, because people will be happy to see that you listen to them gradually)

**Comment by luttje** [comment]
@rayzox57 I found some time to complete this. I have updated the Steam Workshop addon.

I did not take a whole lot of time to test. If you find any bugs/problems/errors, do not hesitate to create a new issue!

Thanks for the suggestion, I hope you enjoy.

**Comment by rayzox57** [comment]
Thank for you fast response ^^

First error (looped) : 
When i take the Golf Club :
![image](https://user-images.githubusercontent.com/27972661/179358598-9ed378c2-4a04-4902-8a2d-ae4b049c3c73.png)
I will fork your addon to try to fix bugs like this ;)

**Comment by rayzox57** [comment]
Oops this error is not you're its my try to transform in addon before ask you x)

**Comment by rayzox57** [comment]
first feedback ;) (ideas) :
- In `minigolf_hole_start` add a new input : `infinite_retry (int)` (in order to be able to replay endlessly even after a disqualification or after finished the hole)

`infinite_retry` values : 
`nil or x = 0 or x > 2` : no retry on disqualification or finished
`x = 1` : no retry on disqualification and retry on finished
`x = 2` : retry in disqualification or finished

`for old maps` : default `nil` because their maps has not their input

- In `minigolf_hole_start` add a new input : `infinite_retry_disqualification_max (int)` (number of max retry when disqualification)

`infinite_retry_disqualification_max` values :
`x <= -1` : infinite
`x > 0` : x retry

`for old maps` : default `0` because their maps has not their input

- In `minigolf_hole_start` add a new input : `infinite_retry_finished_max (int)` (number of max retry when finished hole)

`infinite_retry_finished_max` values :
`x <= -1` : infinite
`x > 0` : x retry

`for old maps` : default `0` because their maps has not their input

**Comment by luttje** [comment]
Good idea! I've created a new issue for that here: #3

I'll have a go at it in the coming week. If you find any bugs I'll fix those then as well.

**Comment by rayzox57** [comment]
![image](https://user-images.githubusercontent.com/27972661/179362081-d65125b8-3f35-4564-8d5e-e929fe917811.png)
I create for you a fgd for all people want to create entities without boring manual add properties x)

**Comment by rayzox57** [comment]
![image](https://user-images.githubusercontent.com/27972661/179363940-be303ac7-77d9-4a6a-86f0-d01b97c8d60e.png)
fgd done

**Comment by luttje** [comment]
Oh very useful! Feel free to make a Pull Request, or simply share the fgd in a new issue and I'll add it to this project the next time I work on it.

Thanks for your help! ☺️
